File: packages/firebase_common/lib/src/firestore_documents/read_status.flutterfire_gen.dart
// coverage:ignore-file
// ignore_for_file: type=lint
// ignore_for_file: unused_element, deprecated_member_use, deprecated_member_use_from_same_package, use_function_type_syntax_for_parameters, unnecessary_const, avoid_init_to_null, invalid_override_different_default_values_named, prefer_expression_function_bodies, annotate_overrides, invalid_annotation_target, unnecessary_question_mark

part of 'read_status.dart';

class ReadReadStatus {
  const ReadReadStatus({
    required this.readStatusId,
    required this.path,
    required this.lastReadAt,
  });

  final String readStatusId;

  final String path;

  final DateTime? lastReadAt;

  factory ReadReadStatus._fromJson(Map<String, dynamic> json) {
    return ReadReadStatus(
      readStatusId: json['readStatusId'] as String,
      path: json['path'] as String,
      lastReadAt: (json['lastReadAt'] as Timestamp?)?.toDate(),
    );
  }

  factory ReadReadStatus.fromDocumentSnapshot(DocumentSnapshot ds) {
    final data = ds.data()! as Map<String, dynamic>;
    return ReadReadStatus._fromJson(<String, dynamic>{
      ...data,
      'readStatusId': ds.id,
      'path': ds.reference.path,
    });
  }
}

class CreateReadStatus {
  const CreateReadStatus();

  Map<String, dynamic> toJson() {
    return {
      'lastReadAt': FieldValue.serverTimestamp(),
    };
  }
}

class UpdateReadStatus {
  const UpdateReadStatus();

  Map<String, dynamic> toJson() {
    return {
      'lastReadAt': FieldValue.serverTimestamp(),
    };
  }
}

class DeleteReadStatus {}

/// Provides a reference to the readStatuses collection for reading.
CollectionReference<ReadReadStatus> readReadStatusCollectionReference({
  required String chatRoomId,
}) =>
    FirebaseFirestore.instance
        .collection('chatRooms')
        .doc(chatRoomId)
        .collection('readStatuses')
        .withConverter<ReadReadStatus>(
          fromFirestore: (ds, _) => ReadReadStatus.fromDocumentSnapshot(ds),
          toFirestore: (_, __) => throw UnimplementedError(),
        );

/// Provides a reference to a readStatus document for reading.
DocumentReference<ReadReadStatus> readReadStatusDocumentReference({
  required String chatRoomId,
  required String readStatusId,
}) =>
    readReadStatusCollectionReference(chatRoomId: chatRoomId).doc(readStatusId);

/// Provides a reference to the readStatuses collection for creating.
CollectionReference<CreateReadStatus> createReadStatusCollectionReference({
  required String chatRoomId,
}) =>
    FirebaseFirestore.instance
        .collection('chatRooms')
        .doc(chatRoomId)
        .collection('readStatuses')
        .withConverter<CreateReadStatus>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (obj, _) => obj.toJson(),
        );

/// Provides a reference to a readStatus document for creating.
DocumentReference<CreateReadStatus> createReadStatusDocumentReference({
  required String chatRoomId,
  required String readStatusId,
}) =>
    createReadStatusCollectionReference(chatRoomId: chatRoomId)
        .doc(readStatusId);

/// Provides a reference to the readStatuses collection for updating.
CollectionReference<UpdateReadStatus> updateReadStatusCollectionReference({
  required String chatRoomId,
}) =>
    FirebaseFirestore.instance
        .collection('chatRooms')
        .doc(chatRoomId)
        .collection('readStatuses')
        .withConverter<UpdateReadStatus>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (obj, _) => obj.toJson(),
        );

/// Provides a reference to a readStatus document for updating.
DocumentReference<UpdateReadStatus> updateReadStatusDocumentReference({
  required String chatRoomId,
  required String readStatusId,
}) =>
    updateReadStatusCollectionReference(chatRoomId: chatRoomId)
        .doc(readStatusId);

/// Provides a reference to the readStatuses collection for deleting.
CollectionReference<DeleteReadStatus> deleteReadStatusCollectionReference({
  required String chatRoomId,
}) =>
    FirebaseFirestore.instance
        .collection('chatRooms')
        .doc(chatRoomId)
        .collection('readStatuses')
        .withConverter<DeleteReadStatus>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (_, __) => throw UnimplementedError(),
        );

/// Provides a reference to a readStatus document for deleting.
DocumentReference<DeleteReadStatus> deleteReadStatusDocumentReference({
  required String chatRoomId,
  required String readStatusId,
}) =>
    deleteReadStatusCollectionReference(chatRoomId: chatRoomId)
        .doc(readStatusId);

/// Manages queries against the readStatuses collection.
class ReadStatusQuery {
  /// Fetches [ReadReadStatus] documents.
  Future<List<ReadReadStatus>> fetchDocuments({
    required String chatRoomId,
    GetOptions? options,
    Query<ReadReadStatus>? Function(Query<ReadReadStatus> query)? queryBuilder,
    int Function(ReadReadStatus lhs, ReadReadStatus rhs)? compare,
  }) async {
    Query<ReadReadStatus> query =
        readReadStatusCollectionReference(chatRoomId: chatRoomId);
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    final qs = await query.get(options);
    final result = qs.docs.map((qds) => qds.data()).toList();
    if (compare != null) {
      result.sort(compare);
    }
    return result;
  }

  /// Subscribes [ReadStatus] documents.
  Stream<List<ReadReadStatus>> subscribeDocuments({
    required String chatRoomId,
    Query<ReadReadStatus>? Function(Query<ReadReadStatus> query)? queryBuilder,
    int Function(ReadReadStatus lhs, ReadReadStatus rhs)? compare,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    Query<ReadReadStatus> query =
        readReadStatusCollectionReference(chatRoomId: chatRoomId);
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    var streamQs =
        query.snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamQs = streamQs.where((qs) => !qs.metadata.hasPendingWrites);
    }
    return streamQs.map((qs) {
      final result = qs.docs.map((qds) => qds.data()).toList();
      if (compare != null) {
        result.sort(compare);
      }
      return result;
    });
  }

  /// Fetches a specific [ReadReadStatus] document.
  Future<ReadReadStatus?> fetchDocument({
    required String chatRoomId,
    required String readStatusId,
    GetOptions? options,
  }) async {
    final ds = await readReadStatusDocumentReference(
      chatRoomId: chatRoomId,
      readStatusId: readStatusId,
    ).get(options);
    return ds.data();
  }

  /// Subscribes a specific [ReadStatus] document.
  Stream<ReadReadStatus?> subscribeDocument({
    required String chatRoomId,
    required String readStatusId,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    var streamDs = readReadStatusDocumentReference(
      chatRoomId: chatRoomId,
      readStatusId: readStatusId,
    ).snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamDs = streamDs.where((ds) => !ds.metadata.hasPendingWrites);
    }
    return streamDs.map((ds) => ds.data());
  }

  /// Adds a [ReadStatus] document.
  Future<DocumentReference<CreateReadStatus>> add({
    required String chatRoomId,
    required CreateReadStatus createReadStatus,
  }) =>
      createReadStatusCollectionReference(chatRoomId: chatRoomId)
          .add(createReadStatus);

  /// Sets a [ReadStatus] document.
  Future<void> set({
    required String chatRoomId,
    required String readStatusId,
    required CreateReadStatus createReadStatus,
    SetOptions? options,
  }) =>
      createReadStatusDocumentReference(
        chatRoomId: chatRoomId,
        readStatusId: readStatusId,
      ).set(createReadStatus, options);

  /// Updates a specific [ReadStatus] document.
  Future<void> update({
    required String chatRoomId,
    required String readStatusId,
    required UpdateReadStatus updateReadStatus,
  }) =>
      updateReadStatusDocumentReference(
        chatRoomId: chatRoomId,
        readStatusId: readStatusId,
      ).update(updateReadStatus.toJson());

  /// Deletes a specific [ReadStatus] document.
  Future<void> delete({
    required String chatRoomId,
    required String readStatusId,
  }) =>
      deleteReadStatusDocumentReference(
        chatRoomId: chatRoomId,
        readStatusId: readStatusId,
      ).delete();
}

File: functions/src/firestore-documents/readStatus.flutterfireGen.ts
import * as admin from 'firebase-admin'
import {
    CollectionReference,
    DocumentData,
    DocumentReference,
    DocumentSnapshot,
    FieldValue,
    Query,
    QueryDocumentSnapshot,
    QuerySnapshot,
    SetOptions,
    Timestamp,
    WriteResult
} from 'firebase-admin/firestore'

export class ReadReadStatus {
    constructor({
        readStatusId,
        path,
        lastReadAt
    }: {
        readStatusId: string
        path: string
        lastReadAt?: Date
    }) {
        this.readStatusId = readStatusId
        this.path = path
        this.lastReadAt = lastReadAt
    }

    readonly readStatusId: string

    readonly path: string

    readonly lastReadAt?: Date

    private static fromJson(json: Record<string, unknown>): ReadReadStatus {
        return new ReadReadStatus({
            readStatusId: json[`readStatusId`] as string,
            path: json[`path`] as string,
            lastReadAt: (json[`lastReadAt`] as Timestamp | undefined)?.toDate()
        })
    }

    static fromDocumentSnapshot(ds: DocumentSnapshot): ReadReadStatus {
        const data = ds.data()!
        const cleanedData: Record<string, unknown> = {}
        for (const [key, value] of Object.entries(data)) {
            cleanedData[key] = value === null ? undefined : value
        }
        return ReadReadStatus.fromJson({
            ...cleanedData,
            readStatusId: ds.id,
            path: ds.ref.path
        })
    }
}

export class CreateReadStatus {
    constructor() {}

    toJson(): Record<string, unknown> {
        return {
            lastReadAt: FieldValue.serverTimestamp()
        }
    }
}

export class UpdateReadStatus {
    constructor() {}

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    toJson(): Record<string, any> {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const json: Record<string, any> = {}
        json[`lastReadAt`] = FieldValue.serverTimestamp()
        return json
    }
}

export class DeleteReadStatus {}

/**
 * A Cloud Firestore object which ignores `undefined` properties.
 */
const db = admin.firestore()
db.settings({ ignoreUndefinedProperties: true })

/**
 * Provides a reference to the readStatuses collection for reading.
 */
export const readReadStatusCollectionReference = ({
    chatRoomId
}: {
    chatRoomId: string
}): CollectionReference<ReadReadStatus> => {
    return db
        .collection(`chatRooms`)
        .doc(chatRoomId)
        .collection(`readStatuses`)
        .withConverter<ReadReadStatus>({
            fromFirestore: (ds: DocumentSnapshot): ReadReadStatus => {
                return ReadReadStatus.fromDocumentSnapshot(ds)
            },
            toFirestore: () => {
                throw new Error(
                    `toFirestore is not implemented for ReadReadStatus`
                )
            }
        })
}

/**
 * Provides a reference to a readStatus document for reading.
 * @param readStatusId - The ID of the readStatus document to read.
 */
export const readReadStatusDocumentReference = ({
    chatRoomId,
    readStatusId
}: {
    chatRoomId: string
    readStatusId: string
}): DocumentReference<ReadReadStatus> =>
    readReadStatusCollectionReference({
        chatRoomId
    }).doc(readStatusId)

/**
 * Provides a reference to the readStatuses collection for creating.
 */
export const createReadStatusCollectionReference = ({
    chatRoomId
}: {
    chatRoomId: string
}): CollectionReference<CreateReadStatus> => {
    return db
        .collection(`chatRooms`)
        .doc(chatRoomId)
        .collection(`readStatuses`)
        .withConverter<CreateReadStatus>({
            fromFirestore: () => {
                throw new Error(
                    `fromFirestore is not implemented for CreateReadStatus`
                )
            },
            toFirestore: (obj: CreateReadStatus): DocumentData => {
                return obj.toJson()
            }
        })
}

/**
 * Provides a reference to a readStatus document for creating.
 * @param readStatusId - The ID of the readStatus document to read.
 */
export const createReadStatusDocumentReference = ({
    chatRoomId,
    readStatusId
}: {
    chatRoomId: string
    readStatusId: string
}): DocumentReference<CreateReadStatus> =>
    createReadStatusCollectionReference({
        chatRoomId
    }).doc(readStatusId)

/**
 * Provides a reference to the readStatuses collection for updating.
 */
export const updateReadStatusCollectionReference = ({
    chatRoomId
}: {
    chatRoomId: string
}): CollectionReference<UpdateReadStatus> => {
    return db
        .collection(`chatRooms`)
        .doc(chatRoomId)
        .collection(`readStatuses`)
        .withConverter<UpdateReadStatus>({
            fromFirestore: () => {
                throw new Error(
                    `fromFirestore is not implemented for UpdateReadStatus`
                )
            },
            toFirestore: (obj: UpdateReadStatus): DocumentData => {
                return obj.toJson()
            }
        })
}

/**
 * Provides a reference to a readStatus document for updating.
 * @param readStatusId - The ID of the readStatus document to read.
 */
export const updateReadStatusDocumentReference = ({
    chatRoomId,
    readStatusId
}: {
    chatRoomId: string
    readStatusId: string
}): DocumentReference<UpdateReadStatus> =>
    updateReadStatusCollectionReference({
        chatRoomId
    }).doc(readStatusId)

/**
 * Provides a reference to the readStatuses collection for deleting.
 */
export const deleteReadStatusCollectionReference = ({
    chatRoomId
}: {
    chatRoomId: string
}): CollectionReference<DeleteReadStatus> => {
    return db
        .collection(`chatRooms`)
        .doc(chatRoomId)
        .collection(`readStatuses`)
        .withConverter<DeleteReadStatus>({
            fromFirestore: () => {
                throw new Error(
                    `fromFirestore is not implemented for DeleteReadStatus`
                )
            },
            toFirestore: (): DocumentData => {
                throw new Error(
                    `toFirestore is not implemented for DeleteReadStatus`
                )
            }
        })
}

/**
 * Provides a reference to a readStatus document for deleting.
 * @param readStatusId - The ID of the readStatus document to read.
 */
export const deleteReadStatusDocumentReference = ({
    chatRoomId,
    readStatusId
}: {
    chatRoomId: string
    readStatusId: string
}): DocumentReference<DeleteReadStatus> =>
    deleteReadStatusCollectionReference({
        chatRoomId
    }).doc(readStatusId)

/**
 * Manages queries against the readStatuses collection.
 */
export class ReadStatusQuery {
    /**
     * Fetches readStatus documents.
     * @param queryBuilder - Function to modify the query.
     * @param compare - Function to sort the results.
     */
    async fetchDocuments({
        chatRoomId,
        queryBuilder,
        compare
    }: {
        chatRoomId: string
        queryBuilder?: (query: Query<ReadReadStatus>) => Query<ReadReadStatus>
        compare?: (lhs: ReadReadStatus, rhs: ReadReadStatus) => number
    }): Promise<ReadReadStatus[]> {
        let query: Query<ReadReadStatus> = readReadStatusCollectionReference({
            chatRoomId
        })
        if (queryBuilder != undefined) {
            query = queryBuilder(query)
        }
        const qs: QuerySnapshot<ReadReadStatus> = await query.get()
        let result = qs.docs.map((qds: QueryDocumentSnapshot<ReadReadStatus>) =>
            qds.data()
        )
        if (compare != undefined) {
            result = result.sort(compare)
        }
        return result
    }

    /**
     * Fetches a specific readStatus document.
     * @param readStatusId - The ID of the readStatus document to fetch.
     */
    async fetchDocument({
        chatRoomId,
        readStatusId
    }: {
        chatRoomId: string
        readStatusId: string
    }): Promise<ReadReadStatus | undefined> {
        const ds = await readReadStatusDocumentReference({
            chatRoomId,
            readStatusId
        }).get()
        return ds.data()
    }

    /**
     * Adds a readStatus document.
     * @param createReadStatus - The readStatus details to add.
     */
    async add({
        chatRoomId,
        createReadStatus
    }: {
        chatRoomId: string
        createReadStatus: CreateReadStatus
    }): Promise<DocumentReference<CreateReadStatus>> {
        return createReadStatusCollectionReference({ chatRoomId }).add(
            createReadStatus
        )
    }

    /**
     * Sets a readStatus document.
     * @param readStatusId - The ID of the readStatus document to set.
     * @param createReadStatus - The readStatus details to set.
     * @param options - Options for the set operation.
     */
    async set({
        chatRoomId,
        readStatusId,
        createReadStatus,
        options
    }: {
        chatRoomId: string
        readStatusId: string
        createReadStatus: CreateReadStatus
        options?: SetOptions
    }): Promise<WriteResult> {
        if (options == undefined) {
            return createReadStatusDocumentReference({
                chatRoomId,
                readStatusId
            }).set(createReadStatus)
        } else {
            return createReadStatusDocumentReference({
                chatRoomId,
                readStatusId
            }).set(createReadStatus, options)
        }
    }

    /**
     * Updates a specific readStatus document.
     * @param readStatusId - The ID of the readStatus document to update.
     * @param updateReadStatus - The details for updating the readStatus.
     */
    async update({
        chatRoomId,
        readStatusId,
        updateReadStatus
    }: {
        chatRoomId: string
        readStatusId: string
        updateReadStatus: UpdateReadStatus
    }): Promise<WriteResult> {
        return updateReadStatusDocumentReference({
            chatRoomId,
            readStatusId
        }).update(updateReadStatus.toJson())
    }

    /**
     * Deletes a specific readStatus document.
     * @param readStatusId - The ID of the readStatus document to delete.
     */
    async delete({
        chatRoomId,
        readStatusId
    }: {
        chatRoomId: string
        readStatusId: string
    }): Promise<WriteResult> {
        return deleteReadStatusDocumentReference({
            chatRoomId,
            readStatusId
        }).delete()
    }
}
